File: frontend/src/app/services/assignment.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Assignment, AssignmentRequest, ReassignmentRequest } from '../models/models';
import { environment } from '../../environments/environment';

@Injectable({ providedIn: 'root' })
export class AssignmentService {
  private url = `${environment.apiBaseUrl}/assignments`;

  constructor(private http: HttpClient) {}

  getByResource(resourceId: string): Observable<Assignment[]> {
    return this.http.get<Assignment[]>(`${this.url}/resource/${resourceId}`);
  }

  getByProject(projectId: string): Observable<Assignment[]> {
    return this.http.get<Assignment[]>(`${this.url}/project/${projectId}`);
  }

  getByBid(bidId: string): Observable<Assignment[]> {
    return this.http.get<Assignment[]>(`${this.url}/bid/${bidId}`);
  }

  create(request: AssignmentRequest): Observable<Assignment> {
    return this.http.post<Assignment>(this.url, request);
  }

  update(id: string, request: AssignmentRequest): Observable<Assignment> {
    return this.http.put<Assignment>(`${this.url}/${id}`, request);
  }

  reassign(request: ReassignmentRequest): Observable<Assignment> {
    return this.http.post<Assignment>(`${this.url}/reassign`, request);
  }

  checkConflicts(resourceId: string, startMonth: number, endMonth: number,
                  allocation: number, excludeId?: string): Observable<{ hasConflict: boolean }> {
    return this.http.get<{ hasConflict: boolean }>(`${this.url}/conflicts`, {
      params: {
        resourceId, startMonth: startMonth.toString(), endMonth: endMonth.toString(),
        allocation: allocation.toString(), ...(excludeId ? { excludeId } : {})
      }
    });
  }

  delete(id: string): Observable<void> {
    return this.http.delete<void>(`${this.url}/${id}`);
  }
}

File: backend/src/main/java/com/resourceplanner/service/AssignmentService.java
package com.resourceplanner.service;

import com.resourceplanner.dto.request.AssignmentRequest;
import com.resourceplanner.dto.request.ReassignmentRequest;
import com.resourceplanner.dto.response.AssignmentResponse;
import java.util.List;
import java.util.UUID;

public interface AssignmentService {
    List<AssignmentResponse> getByResource(UUID resourceId);
    List<AssignmentResponse> getByProject(UUID projectId);
    List<AssignmentResponse> getByBid(UUID bidId);
    AssignmentResponse create(AssignmentRequest request);
    AssignmentResponse update(UUID id, AssignmentRequest request);
    AssignmentResponse reassign(ReassignmentRequest request);
    void delete(UUID id);
    boolean hasConflict(UUID resourceId, int startMonth, int endMonth, int allocation, UUID excludeAssignmentId);
}
